doc/specs/thrift-uri.md

```diff
@@ -0,0 +1,155 @@
+KIND, either express or implied. See the License for the
+specific language governing permissions and limitations
+under the License.
+
+--------------------------------------------------------------------
+-->
+
+# Motivation and use case
+
+Describing the endpoint specifics for a Thrift API today is a purely text-adventure like exercise, which leaves the client side developer end with the tedious task to set up and put together a proper protocol/transport stack. That sometimes even leads to headaches, e.g. if the server requires e.g. framed protocol which might not be obvious. The use of a generally accepted, machine-readable and extensible Thrift URI format to describe client bindings could streamline that process.
+
+# Thrift URI general format
+
+Lets look at the general format:
+
+	"thrift://" <protocol> "/" <transport> ["/" <layer>]* "?" <transport-specific-data>
+
+The *scheme* of all Thrift URIs is "thrift://". Immediately followimng parts are the protocol being used and the endpoint transport.
+
+Optionally, more path segments may be added, each one describing a particular layered transport, e.g. "framed". 
+
+All query data (i.e. the part after the question mark) depend on the endpoint transport in the second segment, examples follow below. All data must be properly URL-encoded.
+
+# Predefined identifiers
+
+Each implementation shall register their implemented formats with the following keys internally:
+
+## Protocols
+|Code|Protocol|
+|-|-|
+|binary|Thrift Binary protocol|
+|compact|Thrift Compact protocol|
+|json|Standard Thrift JSON protocol|
+
+TODO: cover multiplex protocol
+
+## Endpoint Transports
+|Code|Transport|
+|-|-|
+|http|http(s) transport|
+|namedpipes|Named Pipes Transport|
+|pipes|Simple Pipes Transport (i.e. via STDIN,STDOUT)|
+|socket|Socket Transport|
+|tlssocket|TLS Sockets Transport|
+|file|File transport|
+|memory|Memory Buffer Transport|
+
+## Layered Transports
+|Code|Transport|
+|-|-|
+|framed|Framed transport|
+|buffered|Buffered transport|
+|zlib|ZLib transport|
+
+## TODO: multiplex protocol
+....
+
+
+# Extensibility
+
+Consistent with the open and extensible nature of Thrift, the registration mechanism outlined above is intentionally designed to be open to any user-defined protocols and transports. That way, future developments as well as proprietary developments can be covered by the same mechanism. It is also expected, that depending on the implemented set of features, different languages supported by Thrift might support a different set of Thrift URI components.
+
+
+# Transport specific data
+
+
+## http - http(s) transport
+
+The data part consists of the target URL. No other data are allowed.
+
+Examples:
+ * thrift://binary/http?https%3A%2F%2Fuser%3Apass%40example.com%2Fmyservice%3Farg%3Done%26arg%3Dtwo
+
+## namedpipes - Named Pipes Transport
+
+The data part consists of the target pipe, either name only or in full format:
+
+Examples:
+ * thrift://binary/namedpipes?mypipe
```

Comment:
As referenced in https://www.ietf.org/rfc/rfc2396.txt, even though this is not how URI is defined, a common syntax used widely defines it as:

```
<scheme>://<authority><path>?<query>
```

so this will be parsed into:

* `scheme`: `thrift`
* `authority`: `binary`
* `path`: `/namedpipes`
* `query`: `mypipe`

and the `authority` part can be misleading/confusing.

I haven't read that RFC fully carefully yet, but I _think_ that we can add another `/` to make `authority` part empty:

```
thrift:///binary/namedpipes?mypipe
```

or, if that doesn't work, then maybe we should try to define it similar to `mailto:`, without the `//` part followed by `:`.